### ISSUE TITLE
Support customized .terraformrc and credentials.tfrc.json

### ISSUE THREAD

**Issue by chivalryq** [kubevela/terraform-controller#350]
Sometimes users want to use customized `.terraformrc` and `credentials.tfrc.json` files. Here's one use case:

- Use self-hosted JFrog Artifactory to proxy and cache AWS provider resource. These two files are used to point to the self-hosted services and provide credentials.

We should support mount volumes for each file when apply jobs.

Configuration like:

```YAML
apiVersion: terraform.core.oam.dev/v1beta2
kind: Configuration
metadata:
  name: custom-rc
spec:
  cliConfiguationSecretRef:
    name: cli-conf
    namespace: default
  remote: <module>
  ...
```

And the Secret like:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cli-conf
  namespace: default
data:
  terraformrc: |
   ...
  credentials.tfrc.json: |
   ...
```

**Comment by ramekris3163** [comment]
@chivalryq we are going to look into this, we are going to follow the changes done for supporting git ssh credentials as per the commit: https://github.com/kubevela/terraform-controller/commit/841b0d17d9fa00038611f24ee1bfbf76355602b2#diff-1aa78f86e236bbfbdd87a2dd01cfad0cda0964bdfc8ba4a7ba036fd11cbbabb5.

**Comment by chivalryq** [comment]
Hey @ramekris3163. Glad to hear that! You can just start working on this issue. And ping me to review after you done.

**Comment by ramekris3163** [comment]
@chivalryq [pull request](https://github.com/kubevela/terraform-controller/pull/352) for supporting volume mounting of the .terraformrc and credentials.tfrc.json which is required for connecting to a private terraform registry.

ex:

Secret:

`
apiVersion: v1
kind: Secret
metadata:
  name: terraform-creds
  namespace: default
data:
  terraformrc: |
   ...
  credentials.tfrc.json: |
   ...
`

Configuration :

`
apiVersion: terraform.core.oam.dev/v1beta2
kind: Configuration
metadata:
  name: custom-rc
spec:
  TerraformCredentialsSecretReference:
    name: cli-conf
    namespace: default
  remote: <module>
  ...
`

**Comment by ramekris3163** [comment]
@chivalryq we also have some other files or use cases where we want to mount some custom files. 
For ex: we also have a requirement where we can add a credentials helper to the terraform-controller container. So can we make this configurable like:

```yaml
apiVersion: terraform.core.oam.dev/v1beta2
kind: Configuration
metadata:
  name: custom-rc
spec:
  CustomVolumeMount:
    -  name: terraform-creds-helper
       namespace: default
       path: "/root/.terraform.d"
    - name: terrraform-rc
       namespace: default
       path: "/root"
    - name: terraform-credentials
       namespace: deafult
       path: "/root/.terraform.d"
  remote: <module>
  ...
```

**Comment by chivalryq** [comment]
@ramekris3163 About the `CustomVolumeMount` I think it should be discussed more. 
1. Necessity: what is the `terraform-creds-helper` Secret for?
2. The field name `CustomVolumeMount` is confusing if in configuration's spec. At least it shouldn't be a top-level field.
3. The usage of mounted volume defined by `CustomVolumeMount` is not clear. For example, if we specify `gitCredsSecretRef`, we know the secret content should be used in initContainer to for retrieving repo. But it is nit expressed by `CustomVolumeMound`

IMO, more extension for apply/destroy Job should be defined in a more generic way. Like
```yaml
apiVersion: terraform.core.oam.dev/v1beta2
kind: Configuration
metadata:
  name: custom-rc
spec:
  jobPatch:
    apply: 
      spec: 
        volumeMount:
          <volumeMountSpec>
    destroy:
      spec:
        <JobSpec>
```

But this is also a draft.

**Comment by ramekris3163** [comment]
@chivalryq the terraform-creds-helper can be used for authentication instead of credentials.tfrc.json

ex: 

~/.terraformrc : this file configures that we should connect to a private terraform registry url

```
provider_installation {
    direct {
        exclude = ["registry.terraform.io/*/*"]
    }
    network_mirror {
        url = "private-terraform-registry-url"
    }
}
credentials_helper "artifactory" {}
```

~/.terraform.d/credentials.tfrc.json : this file configures the creds for the private terraform registry

```json
{
    "credentials": {
     "private-terraform-registry-url-base-url": {
      "token": "token-value"
     }
    }
   }
```

~/.terraform.d/plugins/terraform-credentials-artifactory : this file can be used to write a script which can retrieve the token by curl or some other way.

```
#!/bin/bash
command="$1"
host="$2"
logfile="credential_helper.log"
if [ "$command" = "get" ]; then
  token=`curl command or some command to retrieve the token`
  if  [ -z "$token" ]; then
    echo "unable to retrieve token" >> $logfile
    echo "{}"
  else
    echo "Token retrieved $token" >> $logfile
    echo "{ \"token\": $token }"
  fi
  exit 0
else
  echo "$host $command" >> $logfile
  exit 1
fi
```  

So the above are the various configurations which can be used to connect to a private terraform registry.

1. The .terraformrc file is used to provide the private terraform registry url.
2. The credentials.tfrc.json is used to provide the token for the private terraform registry.
3. The credentials-helper can be used to provide a script which is run to retrieve the token.


The credentials.tfrc.json and credentials-helper are such that either one can be used for configuring the credentials.

**Comment by ramekris3163** [comment]
@chivalryq https://github.com/kubevela/terraform-controller/pull/352 updated the pull request with the above mentioned changes. Please review and let us know if any other changes are required.

**Comment by chivalryq** [comment]
Fixed by #352
